### ISSUE TITLE
BUG: numpy.nan_to_num failing filling values for infinite values generated by impossible casting

### ISSUE THREAD

**Issue by defoishugo** [numpy/numpy#19861]
Hi,

When filling inf values generated by impossible casting, the *numpy.nan_to_num* fails setting *inf* values to another value.

### Reproducing code example:

```
import numpy as np

array_with_non_float16_values = [
    [839735.3786387637, 83.37],
    [0.0, 839735.3786387637],
    [123, 95.87]
]

dtype = [
    ("a", "float16"),
    ("b", "float16")
]

np_array = np.array(array_with_non_float16_values, dtype=dtype)
np_array_without_nan = np.nan_to_num(np_array, nan=0, posinf=0, neginf=0)

print(np_array_without_nan)
```

### Output obtained:

```
> python test.py
[[(  inf,   inf) ( 83.4,  83.4)]
 [(  0. ,   0. ) (  inf,   inf)]
 [(123. , 123. ) ( 95.9,  95.9)]]
```

### Output desired:

```
> python test.py
[[(  0. ,   0. ) ( 83.4,  83.4)]
 [(  0. ,   0. ) (  0. ,   0. )]
 [(123. , 123. ) ( 95.9,  95.9)]]
```

### NumPy/Python version information:

1.21.1 3.8.10 | packaged by conda-forge | (default, May 11 2021, 06:25:23) [MSC v.1916 64 bit (AMD64)]

### Comments:

I would like to work on resolving this issue.
I can do a pull-request as soon as possible.

This issue is blocking some of my projects, so I really need to provide a fix.

**Comment by seberg** [comment]
@defoishugo sorry, I don't understand what you mean?  The issue I see is that the structured dtype is unsupported by `nan_to_num`?  Which I guess is a bug, the function should probably just fail at least.

**Comment by defoishugo** [comment]
> @defoishugo sorry, I don't understand what you mean? The issue I see is that the structured dtype is unsupported by `nan_to_num`? Which I guess is a bug, the function should probably just fail at least.

@seberg You are right, this is related to the structured dtype, since this code works pretty well filling the *inf* values:

```
import numpy as np

array_with_non_float16_values = [839735.3786387637, 83.37, 0.0, 839735.3786387637]

np_array = np.array(array_with_non_float16_values, dtype="float16")
np_array_without_nan = np.nan_to_num(np_array, nan=0, posinf=0, neginf=0)

print(np_array_without_nan)
```

So the issue is indeed about handling the structured dtype.
Should we consider that as a bug of an enhancement?

**Comment by seberg** [comment]
Honestly, :shrug:...

I am not immediately sure what should ideally happen here.  My first guess would be deprecation/an error, but of course you can also argue that structured dtypes should be a first-class citizen here.
The function works with "elements" and "elements" is the whole structure.  And `isinf`, `isnan` can't be defined for structured dtypes – at least not on the "per element" level that the function is currently implemented at.

The no-op for integers can be seen as an optimization since `isinf` is always false anyway.

**Comment by defoishugo** [comment]
> 
> 
> Honestly, 🤷...
> 
> I am not immediately sure what should ideally happen here. My first guess would be deprecation/an error, but of course you can also argue that structured dtypes should be a first-class citizen here.
> The function works with "elements" and "elements" is the whole structure. And `isinf`, `isnan` can't be defined for structured dtypes – at least not on the "per element" level that the function is currently implemented at.
> 
> The no-op for integers can be seen as an optimization since `isinf` is always false anyway.

# The solution I found for now

For now, the only solution I found is using pandas to fill the values...

```
# Not working
np_array_without_nan = np.nan_to_num(np_array, nan=0, posinf=0, neginf=0)

# Working
pandas_df = pd.DataFrame(data=np_array, dtype="float16")
pandas_df.replace([np.inf, -np.inf], 0, inplace=True)
print(pandas_df)
```

# The numpy solution?

I totally understand your technical point of view. Right now, we don't have any way to deal with structured dtypes using `isinf` or `isnan` so it's not managable to **easily** implement a fix. 

First of all, we should maybe implement the structured dtype support for `isinf` and `isnan`. Maybe should we create a different algorithm to deal with dtypes? Do you think this is something to put on the table or not for now?

If this is not possible to design a structured dtype support, maybe should we create a new function for filling *nan*/*inf* values?

Of course, if I'm the only one of the community using ndarrays with stuctured dtypes, I totally understand that we'll not implement that. But if you feel any needs on the community side, maybe we should think to that!

**Comment by seberg** [comment]
If you know your dtype is structured (or even if not). It should be easy to copy the numpy function and make it work recursively for sturctured dtypes.

**Comment by defoishugo** [comment]
@seberg If I do any pull-request for implementing this auto-behaviour without any user custom code, will you accept it?

**Comment by defoishugo** [comment]
@seberg should I close this issue?

**Comment by seberg** [comment]
It seems like a valid issue.  I just don't have a huge opinion on how we should approach it, but then, I don't use them a lot and don't have a good feel for what the expectations should be.
There could be a decision that all "ufunc likes" should "unpack" structured dtypes (but ufuncs cannot do that, at least not yet); but I am not really aware that this is the way/prior art (so maybe such a decision already exists elsewhere in NumPy?).  (And I guess this is a ufunc-like or could be a ufunc, similar to `np.clip`)
